### PR TITLE
#1735 Add variant manager for data interface

### LINKED ISSUE #1735: Add thread safe management of data interface
To avoid reading multiple times the same IIDM file (very time consuming), we use in cpp11 the powsybl feature that allows to create variants for network.
We also add protection against concurrency access

### PR DESCRIPTION
**Pull request by lecourtoisflo**
Also add protection against concurrent access

closes #1735

### CODE REVIEW

File: dynawo/sources/API/JOB/JOBOutputsEntry.h [hunk 1]
```diff
@@ -39,6 +39,25 @@ namespace job {
...
  public:
+  /// @brief Destructor
+  ~OutputsEntry();
+
+  /// @brief Default constructor
+  OutputsEntry();
+
+  /**
+   * @brief Copy constructor
+   * @param other originalto copy
```
Comment (rosiereflo) [inline dynawo/sources/API/JOB/JOBOutputsEntry.h:50]: typo

File: dynawo/sources/API/JOB/test/TestJobEntry.cpp [hunk 2]
```diff
@@ -53,6 +55,13 @@ TEST(APIJOBTest, testJobEntry) {
   ASSERT_EQ(job->getSolverEntry(), solver);
   ASSERT_EQ(job->getSimulationEntry(), simulation);
   ASSERT_EQ(job->getOutputsEntry(), outputs);
+
+  boost::shared_ptr<JobEntry> job_bis = DYN::clone(job);
```
Comment (rosiereflo) [inline dynawo/sources/API/JOB/test/TestJobEntry.cpp:59]: This only test constructor by copy and not assignment (same goes for all tests)

File: dynawo/sources/Common/DYNTimer.cpp [hunk 3]
```diff
@@ -49,8 +58,15 @@ Timers::add(const std::string& name, const double& time) {
...
-  nbAppels_[name] += 1;
+#ifdef LANG_CXX11
+  std::stringstream ss;
+  ss << std::this_thread::get_id() << "_" << name;
+  std::string name_formatted = ss.str();
+#else
+  std::string name_formatted = name;
+#endif
+  timers_[name_formatted] += time;
+  nbAppels_[name_formatted] += 1;
```
Comment (rosiereflo) [inline dynawo/sources/Common/DYNTimer.cpp:69]: Was this tested? What does the output looks like (just by curiosity)?

File: dynawo/sources/Modeler/DataInterface/DYNDataInterface.h [hunk 4]
```diff
@@ -38,6 +38,18 @@ class DataInterface {
    */
   virtual ~DataInterface() { }
 
+  /**
+   * @brief Determines if variant mangement is supported
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNDataInterface.h:42]: typo management

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
+//
+// Copyright (c) 2015-2020, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp:2]: 2021

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp [hunk 6]
```diff
@@ -102,21 +113,35 @@ DataInterfaceIIDM::build(std::string iidmFilePath) {
...
-    powsybl::iidm::Network networkIIDM = powsybl::iidm::Network::readXml(boost::filesystem::path(iidmFilePath), options);
+    auto networkIIDM = boost::make_shared<powsybl::iidm::Network>(powsybl::iidm::Network::readXml(boost::filesystem::path(iidmFilePath), options));
 
-    data.reset(new DataInterfaceIIDM(std::move(networkIIDM)));
+    if (nbVariants > 1) {
+      auto& manager = networkIIDM->getVariantManager();
+      manager.allowVariantMultiThreadAccess(true);
+      for (unsigned int i = 0; i < nbVariants; ++i) {
+        const std::string& variantName = std::to_string(i);
+        constexpr bool overwrite = true;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:125]: Please create this one outside of the loop

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp [hunk 7]
```diff
@@ -1637,19 +1640,194 @@ TEST(DataInterfaceIIDMTest, testImportError) {
...
+  ASSERT_EQ(outputNetwork.getId(), inputNetwork.getId());
+  ASSERT_EQ(outputNetwork.getId(), network->getId());
+  ASSERT_EQ(inputNetwork.getId(), network->getId());
+}
+
+TEST(DataInterfaceIIDMTest, testClone) {
+  auto network = createNodeBreakerNetworkIIDM();
+
+  shared_ptr<DataInterfaceIIDM> data = createDataItfFromNetwork(network);
+  shared_ptr<DataInterfaceIIDM> data2 = boost::dynamic_pointer_cast<DataInterfaceIIDM>(data->clone());
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:1663]: only test copy constructor and not assignment

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNLoadInterfaceIIDM.h [hunk 8]
```diff
@@ -143,6 +143,12 @@ class LoadInterfaceIIDM : public LoadInterface, public InjectorInterfaceIIDM {
     country_ = country;
   }
 
+#if _DEBUG_
+  powsybl::iidm::Load& getUnderlyingLoad() {
+    return loadIIDM_;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNLoadInterfaceIIDM.h:148]: Discussion ongoing with powsybl team on this one

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/API/JOB/JOBOutputsEntry.h (hunk 1 above)
Comment (lecourtoisflo) [reply dynawo/sources/API/JOB/JOBOutputsEntry.h:50]: OK

File: dynawo/sources/API/JOB/test/TestJobEntry.cpp (hunk 2 above)
Comment (lecourtoisflo) [reply dynawo/sources/API/JOB/test/TestJobEntry.cpp:59]: OK

File: dynawo/sources/Common/DYNTimer.cpp (hunk 3 above)
Comment (lecourtoisflo) [reply dynawo/sources/Common/DYNTimer.cpp:69]: I didn't change this part so I didn't tested it

File: dynawo/sources/Modeler/DataInterface/DYNDataInterface.h (hunk 4 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/DYNDataInterface.h:42]: OK

File: dynawo/sources/Common/DYNTimer.cpp (hunk 3 above)
Comment (rosiereflo) [reply dynawo/sources/Common/DYNTimer.cpp:69]: I mean launching a SA in debug mode and taking a look at the output

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp [hunk 9]
```diff
@@ -313,10 +314,10 @@ DataInterfaceIIDM::getBusName(const std::string& componentName, const std::strin
 void
 DataInterfaceIIDM::initFromIIDM() {
   // create network interface
-  network_.reset(new NetworkInterfaceIIDM(networkIIDM_));
+  network_.reset(new NetworkInterfaceIIDM(*networkIIDM_));
 
-  IIDM::Contains<IIDM::Substation>::iterator itSubstation = networkIIDM_.substations().begin();
-  for (; itSubstation != networkIIDM_.substations().end(); ++itSubstation) {
+  IIDM::Contains<IIDM::Substation>::iterator itSubstation = networkIIDM_->substations().begin();
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp:319]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp (hunk 5 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp:2]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+//
+
+/**
+ * @file Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp
+ * @brief Unit tests for Criteria classes
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp:15]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+  ASSERT_TRUE(data);
+  ASSERT_NE(data, data2);
+
+  ASSERT_EQ(data->getNetworkIIDM().id(), data2->getNetworkIIDM().id());
+
+  boost::shared_ptr<NetworkInterfaceIIDM> network_interface = boost::dynamic_pointer_cast<NetworkInterfaceIIDM>(data->getNetwork());
+  boost::shared_ptr<NetworkInterfaceIIDM> network_interface2 = boost::dynamic_pointer_cast<NetworkInterfaceIIDM>(data2->getNetwork());
+  ASSERT_NE(network_interface, network_interface2);
+  ASSERT_EQ(network_interface->getLines().size(), network_interface2->getLines().size());
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp:124]: This still test that the number of lines is the same

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp [hunk 12]
```diff
@@ -75,17 +75,28 @@ using boost::dynamic_pointer_cast;
...
 
 DataInterfaceIIDM::~DataInterfaceIIDM() {
 }
 
+void
+DataInterfaceIIDM::loadExtensions(const std::vector<std::string>& paths) {
+  std::unique_lock<std::mutex> lock(loadExtensionMutex_);
+  for (const auto& path : paths) {
+    std::regex fileRegex(stdcxx::format(".*libiidm-ext-.*\\%1%", boost::dll::shared_library::suffix().string()));
+    powsybl::iidm::ExtensionProviders<powsybl::iidm::converter::xml::ExtensionXmlSerializer>::getInstance().loadExtensions(path, fileRegex);
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:94]: There is a lock so we ensure that there is no concurrency access. As for the inside of the function, we were already calling it from multiple threads (Call to data interface build function)

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp (hunk 6 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:125]: OK

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp [hunk 13]
```diff
@@ -1082,6 +1101,27 @@ DataInterfaceIIDM::getStaticParameterBoolValue(const std::string& staticID, cons
   return findComponent(staticID)->getStaticParameterValue<bool>(refOrigName);
 }
 
+void
+DataInterfaceIIDM::copy(const DataInterfaceIIDM& other) {
+  networkIIDM_  = other.networkIIDM_;  // No clone here because iidm network is not copyable
+  // Criterias are not copied and must be initialized again
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:1107]: In simulation, when we initialize from dyd: Simulation::loadDynamicData(),  data_->configureCriteria(criteriaCollection_);

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp (hunk 7 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:1663]: Aim is to test the clone, not the operators. Should I put the operators in private then ?

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp:124]: This number is 0 for both

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:94]: ok

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp (hunk 7 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:1663]: any method that exists need to be tester properly

Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:1663]: If they are private they cannot be tested

Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:1663]: either they are used and needs to be tested either they are useless and should be removed no?

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp [hunk 14]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2015-2020, RTE (http://www.rte-france.com)
+// Copyright (c) 2015-2021, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:2]: Please do not modify copyright of existing files

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp (hunk 14 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:2]: The I don't get why there is a range

Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:2]: there is a range for files that were created before the open sourcing of the code (between 2015 and 2019). For new files no need to have a range, current year is enough, but sometimes the range is kept (due to a copy paste for example) and it makes no harm. 
However we should not need to modify the copyright of existing files.

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNLoadInterfaceIIDM.h (hunk 8 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNLoadInterfaceIIDM.h]: Removed

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp (hunk 14 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestDataInterface.cpp:2]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/test/TestDataInterfaceIIDM.cpp:124]: OK

File: dynawo/sources/Common/DYNTimer.cpp (hunk 3 above)
Comment (lecourtoisflo) [reply dynawo/sources/Common/DYNTimer.cpp:69]: TIMER[140258212005952_evalJt_evalDerivatives] = 0.073168 secondes en 369 appels
TIMER[140258212005952_evalJt_initBusDerivatives] = 0.015188 secondes en 369 appels
TIMER[140258212005952_zzz filling] = 0.190309 secondes en 5968 appels
TIMER[140258212005952_zzz reading] = 0.575743 secondes en 5968 appels

**Review by rosiereflo — APPROVED** [review]

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp [hunk 15]
```diff
@@ -256,40 +281,40 @@ DataInterfaceIIDM::initFromIIDM() {
...
+  for (auto& hvdcLineIIDM : networkIIDM_->getHvdcLines()) {
+    auto hvdc = importHvdcLine(hvdcLineIIDM);
     network_->addHvdcLine(hvdc);
     components_[hvdc->getID()] = hvdc;
   }
   DYNErrorQueue::instance().flush();
 }
 
-shared_ptr<VoltageLevelInterface>
+shared_ptr<VoltageLevelInterfaceIIDM>
```
Comment (gautierbureau) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:317]: why now use the child class and not the interface ?

**Review by gautierbureau — APPROVED** [review]

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp (hunk 15 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:317]: as it makes useless dynamic cast whereas we know at this point the accurate class

Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:317]: so this is not a change that was required by this dev ? and why not do it for the old implem ?

Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNDataInterfaceIIDM.cpp:317]: It is used to avoidsome dynamic cast in this dev, the old  implementation is supposed to be deleted at short/mid term so no there is no need to report code quality modifications into it

**Pull request merged**